sdks/python/apache_beam/runners/worker/bundle_processor.py

```diff
@@ -823,6 +825,38 @@ def only_element(iterable):
   return element
 
 
+def _extract_py_version(capability):
+  return capability[len(_PY_SDK_CAPABILITY_PREFIX):].split("_sdk:")[0]
+
+
+def _verify_descriptor_created_in_a_compatible_env(process_bundle_descriptor):
+  # type: beam_fn_api_pb2.ProcessBundleDescriptor -> None
+
+  py_envs_at_submission = set()
+  for _, env in process_bundle_descriptor.environments.items():
```

Comment:
Rather than enumerating over all environments, just enumerate over the environments associated with the transforms of this bundle descriptor such that we can require an exact match.

Reply:
thanks for feedback. i am also thinking to stop checking as soon as one environment successfully matched, and also considering doing the matches at most once per process. let me know if you think it's over-optimizing.

Reply:
> thanks for feedback. i am also thinking to stop checking as soon as one environment successfully matched, 

We should require that all the environments (referenced in this bundles transforms) match. 

> and also considering doing the matches at most once in process lifetime. let me know if you think it's over-optimizing.

We need to repeatedly raise the error to get the pipeline to fail. This happens once per (already expensive and rare) bundle processor initialization, so I wouldn't worry about trying to limit it to once per process (though a stronger case could be made about warnings about package version mismatch).

Reply:
> Rather than enumerating over all environments, just enumerate over the environments associated with the transforms of this bundle descriptor such that we can require an exact match.

Made the changes, logic is simpler now. However: looks like Dataflow's runner v2 is not setting environment_id field on transforms in ProcessBundleDescriptor. Perhaps, due to an implicit assumption that there is only 1 environment per SDK harness? I also noticed that UW harness starts with a mapping of {SDK_harness_id: environment_id}. However this mapping is not visible to SDK harness. There is still a list of environments in PBD.  
 
 Works with PortableRunner.

Reply:
>  There is still a list of environments in PBD.

... and it includes an "empty" environment:

```
environments {
  key: ""
  value {
  }
}
environments {
  key: "ref_Environment_default_environment_1"
  value {
    urn: "beam:env:docker:v1"
    payload: "\n apache/beam_python3.9_sdk:2.44.0"
    capabilities: "beam:coder:bytes:v1"
    capabilities: "beam:coder:string_utf8:v1"
    capabilities: "beam:coder:kv:v1"
    capabilities: "beam:coder:bool:v1"
    capabilities: "beam:coder:varint:v1"
    capabilities: "beam:coder:double:v1"
    capabilities: "beam:coder:iterable:v1"
    capabilities: "beam:coder:timer:v1"
    capabilities: "beam:coder:interval_window:v1"
    capabilities: "beam:coder:length_prefix:v1"
    capabilities: "beam:coder:global_window:v1"
    capabilities: "beam:coder:windowed_value:v1"
    capabilities: "beam:coder:param_windowed_value:v1"
    capabilities: "beam:coder:state_backed_iterable:v1"
    capabilities: "beam:coder:custom_window:v1"
    capabilities: "beam:coder:row:v1"
    capabilities: "beam:coder:sharded_key:v1"
    capabilities: "beam:coder:nullable:v1"
    capabilities: "beam:protocol:progress_reporting:v0"
    capabilities: "beam:protocol:harness_monitoring_infos:v1"
    capabilities: "beam:protocol:worker_status:v1"
    capabilities: "beam:combinefn:packed_python:v1"
    capabilities: "beam:version:sdk_base:apache/beam_python3.8_sdk:2.46.0.dev"
    capabilities: "beam:transform:sdf_truncate_sized_restrictions:v1"
    capabilities: "beam:transform:to_string:v1"
    capabilities: "beam:protocol:sibling_workers:v1"
    dependencies {
      type_urn: "beam:artifact:type:url:v1"
      type_payload: "\nlgs://clouddfe-valentyn/beamapp-valentyn-0207064357-172319-kxdzeic6.1675752237.172464/dataflow_python_sdk.tar"
      role_urn: "beam:artifact:role:staging_to:v1"
      role_payload: "\n\027dataflow_python_sdk.tar"
    }
    dependencies {
      type_urn: "beam:artifact:type:url:v1"
      type_payload: "\nigs://clouddfe-valentyn/beamapp-valentyn-0207064357-172319-kxdzeic6.1675752237.172464/pickled_main_session"
      role_urn: "beam:artifact:role:staging_to:v1"
      role_payload: "\n\024pickled_main_session"
    }
  }
}
```

Reply:
Interesting. I can take a look on the UW side.